lib/ziti-tunnel-cbs/ziti_dns.c

```diff
@@ -639,20 +655,23 @@ static void proxy_domain_req(struct dns_req *req, dns_domain_t *domain) {
     dns_question *q = req->msg.question[0];
     if (domain->resolv_proxy != NULL && (q->type == NS_T_MX || q->type == NS_T_SRV || q->type == NS_T_TXT)) {
         size_t jsonlen;
-        char *json = dns_message_to_json(&req->msg, MODEL_JSON_COMPACT, &jsonlen);
-        ZITI_LOG(DEBUG, "writing proxy resolve req[%04x]: %s", req->id, json);
+        struct proxy_dns_req_wr_s *wr = calloc(1, sizeof(struct proxy_dns_req_wr_s));
+        wr->req = req;
+        wr->json = dns_message_to_json(&req->msg, MODEL_JSON_COMPACT, &jsonlen);
+        ZITI_LOG(DEBUG, "writing proxy resolve req[%04x]: %s", req->id, wr->json);
 
         // intercept_resolve_connect above can quick-fail if context does not have a valid API session
         // in that case resolve_proxy connection will be in Closed state and write will fail
-        int rc = ziti_write(domain->resolv_proxy, json, jsonlen, on_proxy_write, json);
+        int rc = ziti_write(domain->resolv_proxy, wr->json, jsonlen, on_proxy_write, wr);
         if (rc == ZITI_OK) {
             return;
         }
 
         ZITI_LOG(WARN, "failed to write proxy resolve request[%04x]: %s", req->id, ziti_errorstr(rc));
         ziti_close(domain->resolv_proxy, NULL);
         domain->resolv_proxy = NULL;
-        free(json);
+        free(wr->json);
```

Comment:
this sure seems like a double free to me, but it must not be somehow. wouldn't the on_proxy_write free these?

Reply:
The write callback won’t get called if ziti_write returns error.

Reply:
aaaahhhhh... ok. that makes more sense but wasn't obvious.